### ISSUE TITLE
Very bad performance with gulp-uglify

### ISSUE THREAD

**Issue by perlun** [terinjokes/gulp-uglify#158]
I have seen very slow speeds with gulp-uglify recently, both locally (OS X, `node v4.1.1` and on Travis). This is the output from the latter:

```
Pre-installed Node.js versions
v0.10.36

...

[14:27:21] Requiring external module coffee-script/register
[14:27:22] Using gulpfile ~/build/ecraft/indirect-purchasing-backoffice-app/gulpfile.coffee
[14:27:22] Starting 'clean'...
[14:27:22] Finished 'clean' after 11 ms
[14:27:22] Starting 'dist'...
[14:27:22] Starting 'app.coffee'...
[14:27:22] Starting 'components-coffee'...
[14:27:23] Starting 'components-templates'...
[14:27:23] Starting 'controllers'...
[14:27:23] Starting 'fonts'...
[14:27:23] Starting 'images'...
[14:27:23] Starting 'route-templates'...
[14:27:23] Starting 'misc-templates'...
[14:27:23] Starting 'routes-coffee'...
[14:27:23] Starting 'routes-controllers-coffee'...
[14:27:23] Starting 'initializers'...
[14:27:23] Starting 'lib'...
[14:27:23] Starting 'models'...
[14:27:23] Starting 'router.coffee'...
[14:27:23] Starting 'serializers'...
[14:27:23] Starting 'services'...
[14:27:23] Starting 'styles'...
[14:27:23] Starting 'transforms'...
[14:27:23] Starting 'vendor'...
[14:27:23] Starting 'helpers'...
[14:27:23] Finished 'dist' after 91 ms
[14:27:23] Finished 'misc-templates' after 42 ms
[14:27:23] Finished 'controllers' after 445 ms
[14:27:23] Finished 'images' after 442 ms
[14:27:23] Finished 'serializers' after 430 ms
[14:27:23] Finished 'app.coffee' after 512 ms
[14:27:23] Finished 'router.coffee' after 435 ms
[14:27:23] Finished 'services' after 703 ms
[14:27:24] Finished 'initializers' after 1.74 s
[14:27:24] Finished 'transforms' after 1.75 s
[14:27:25] Finished 'helpers' after 2.03 s
[14:27:25] Finished 'lib' after 2.12 s
[14:27:25] Finished 'fonts' after 2.25 s
[14:27:25] Finished 'routes-coffee' after 2.41 s
[14:27:25] Finished 'styles' after 2.43 s
[14:27:25] Finished 'route-templates' after 2.52 s
[14:27:25] Starting 'templates'...
[14:27:25] Finished 'templates' after 9.33 μs
[14:27:25] Finished 'models' after 2.71 s
[14:27:25] Finished 'routes-controllers-coffee' after 2.89 s
[14:27:25] Starting 'routes'...
[14:27:25] Finished 'routes' after 8.32 μs
[14:27:26] Finished 'vendor' after 3.62 s
[14:27:26] Finished 'components-templates' after 3.7 s
[14:27:26] Finished 'components-coffee' after 3.72 s
[14:27:26] Starting 'components'...
[14:27:26] Finished 'components' after 4.65 μs
[14:27:26] Starting 'index.html'...
[14:27:26] Finished 'index.html' after 50 ms
[14:27:26] Starting 'build'...
[14:27:26] Finished 'build' after 8.59 μs
[14:27:26] Starting 'start-package'...
[14:27:26] Starting 'package'...
[14:27:26] Finished 'start-package' after 12 ms
[14:31:08] Finished 'package' after 3.68 min
```

3.68 **minutes** to perform the uglification.

Here is the code for the `package` task (in CoffeeScript syntax):

```coffee
uglify = require('gulp-uglify')

gulp.task 'package', () ->
  gulp.src appSources
    .pipe concat 'app.js'
    .pipe gulp.dest './dist'

  gulp.src './ui/fonts/*'
    .pipe gulp.dest './dist/fonts'

  gulp.src './ui/images/*'
    .pipe gulp.dest './dist/images'

  gulp.src './ui/vendor/**'
    .pipe gulp.dest './dist/vendor'

  gulp.src vendorSources
    .pipe concat 'vendor.js'
    .pipe uglify(
      mangle: true
    )
    .pipe gulp.dest './dist'
```

If I remove the `.pipe uglify` part, the processing is loads faster, so I'm very certain that `gulp-uglify` is causing it somehow. Any ideas? (using v1.5.1, tried upgrading to the latest to see if it would help but no difference w/ this version compared to previous ones)

**Comment by terinjokes** [comment]
@perlun Do you encounter the same performance issue calling UglifyJS2 from the command line? I've seen issues where UglifyJS causes GC pauses, and want to rule it out before investigating farther.

---

A nit-pick: Each `gulp.src` starts a stream, but you only return the last stream (which is how Gulp determines if tasks are done or not). Not a problem here, as the UglifyJS task likely takes longer than your moves (even without any bugs!), but something you might want to keep in mind.

**Comment by perlun** [comment]
Good idea, I tried the following:

```
$ time node_modules/gulp-uglify/node_modules/uglify-js/bin/uglifyjs --compress --mangle -- ./ui/vendor/jquery-1*.js ./ui/vendor/lodash-*.js ./ui/vendor/jquery.*.js ./ui/vendor/jszip.js ./ui/vendor/ember-1*.js ./ui/vendor/loader-*.js ./ui/vendor/ember-resolver-*.js ./ui/vendor/bootstrap-datepicker.js ./ui/vendor/**/*.js

...

real	2m48.377s
user	2m33.172s
sys	0m4.416s
```

I tried disabling the `--mangle` and `--compress` options, and it was now incredibly fast. Tried with only `--compress`, seemed slow (aborted after 52s). Then tried with only `--mangle`, which was fast (5s, which is OK in this case).

So, in essence, the problem is with `uglify-js`. I will submit a bug request there. Thanks!

**Comment by terinjokes** [comment]
@perlun If you can link the bug that you created (or, if you've found a similar one already, commented on), it would be great if you can leave a comment here with a link.

**Comment by perlun** [comment]
Sure: https://github.com/mishoo/UglifyJS2/issues/917

**Comment by perlun** [comment]
Interesting @terinjokes. I tried applying some of your suggestions, i.e. return an array of `gulp.src` objects instead of just the last one. And you know what?

It made things _extremely_ fast. I just had a 47s time (with compression disabled), possibly caused by some .js file already been minified.

I fixed the method to return an array. And voila, it now completes in _milliseconds_. I cannot believe my eyes actually.

Lesson learned: do things right, it will pay off in the long run. :smile:
